src/compiler/sys.ts

```diff
@@ -232,6 +231,83 @@ namespace ts {
             const platform: string = _os.platform();
             const useCaseSensitiveFileNames = isFileSystemCaseSensitive();
 
+            function fsWatchFile(fileName: string, callback: FileWatcherCallback, pollingInterval?: number): FileWatcher {
+                _fs.watchFile(fileName, { persistent: true, interval: pollingInterval || 250 }, fileChanged);
+                return {
+                    close: () => _fs.unwatchFile(fileName, fileChanged)
+                };
+
+                function fileChanged(curr: any, prev: any) {
+                    const isCurrZero = +curr.mtime === 0;
+                    const isPrevZero = +prev.mtime === 0;
+                    const created = !isCurrZero && isPrevZero;
+                    const deleted = isCurrZero && !isPrevZero;
+
+                    const eventKind = created
+                        ? FileWatcherEventKind.Created
+                        : deleted
+                            ? FileWatcherEventKind.Deleted
+                            : FileWatcherEventKind.Changed;
+
+                    if (eventKind === FileWatcherEventKind.Changed && +curr.mtime <= +prev.mtime) {
+                        return;
+                    }
+
+                    callback(fileName, eventKind);
+                }
+            }
+
+            function fsWatchDirectory(directoryName: string, callback: (eventName: string, relativeFileName: string) => void, recursive?: boolean): FileWatcher {
+                // Node 4.0 `fs.watch` function supports the "recursive" option on both OSX and Windows
+                // (ref: https://github.com/nodejs/node/pull/2649 and https://github.com/Microsoft/TypeScript/issues/4643)
+                let options: any;

```

Comment:
Can't you declare this inside `watchPresentDirectory()`?

Reply:
No because it is used throughout the span of directory watcher.. The watcher can change from presentdirectory watcher to missing directory watcher depending on directory added/deleted

Reply:
Are you saying the `options` object's fields are mutated, so it matters if there is only one of them vs one per call to `watchPresentDirectory`?